src/qibolab/_core/instruments/qcs.py

```diff
@@ -0,0 +1,343 @@
+        # return qcs.ArbitraryEnvelope(
+        #    times=np.linspace(0, 1, num_samples), amplitudes=raw_envelope
+        # )
+        raise Exception("Envelope not supported")
+
+
+def generate_qcs_rfwaveform(
+    duration: Union[float, qcs.Scalar],
+    envelope: Envelope,
+    amplitude: Union[float, qcs.Scalar],
+    frequency: Union[float, qcs.Scalar],
+    phase: Union[float, qcs.Scalar],
+) -> qcs.RFWaveform:
+
+    return qcs.RFWaveform(
+        duration=duration,
+        envelope=generate_qcs_envelope(envelope),
+        amplitude=amplitude,
+        rf_frequency=frequency,
+        instantaneous_phase=phase,
+    )
+
+
+class KeysightQCS(Controller):
+    """Driver for interacting with QCS controller server."""
+
+    bounds: str = "qcs/bounds"
+
+    # Map of QCS virtual channels to QCS physical channels
+    qcs_channel_map: qcs.ChannelMapper
+    # Map of Qibolab channel IDs to QCS virtual channels
+    virtual_channel_map: dict[ChannelId, qcs.Channels]
+    # Map of QCS virtual acquisition channels to QCS state classifiers
+    classifier_map: Optional[dict[qcs.Channels, qcs.MinimumDistanceClassifier]] = None
+
+    def connect(self):
+        self.backend = qcs.HclBackend(self.qcs_channel_map, hw_demod=True)
+        self.backend.is_system_ready()
+
+    @property
+    def sampling_rate(self):
+        return qcs.SAMPLE_RATES[qcs.InstrumentEnum.M5300AWG] / 1e9
+
+    def create_program(
+        self,
+        sequence: PulseSequence,
+        configs: dict[str, Config],
+        sweepers: list[ParallelSweepers],
+        num_shots: int,
+    ) -> qcs.Program:
+        program = qcs.Program()
+
+        # SWEEPER MANAGEMENT
+        # Mapper for pulses that are controlled by a sweeper and the parameter to be swept
+        sweeper_pulse_map: dict[PulseId, dict[str, qcs.Scalar]] = {}
+        # Mapper for channels with frequency controlled by a sweeper
+        sweeper_channel_map: dict[ChannelId, qcs.Scalar] = {}
+        hw_demod = True
+
+        for idx, parallel_sweeper in enumerate(sweepers):
+            sweep_values = []
+            sweep_variables = []
+
+            for idx2, sweeper in enumerate(parallel_sweeper):
+
+                qcs_variable = qcs.Scalar(
+                    name=f"V{idx}_{idx2}", value=sweeper.values[0], dtype=float
+                )
+
+                if sweeper.parameter is Parameter.frequency:
+                    for channel_id in sweeper.channels:
+                        sweeper_channel_map[channel_id] = qcs_variable
+                        # Ignore hardware sweeping if frequency on readout is swept
+                        # TODO: Find better way of determining if channel is readout probe
+                        if (
+                            "readout" in self.virtual_channel_map[channel_id].name
+                            and hw_demod
+                        ):
+                            program = program.n_shots(num_shots)
```

Comment:
So, this is a quirk of the QCS, and maybe applicable to guiding the logic of https://github.com/qiboteam/qibolab/issues/917.

```python
program = program.sweep(sweep_values, sweep_variables).n_shots(num_shots).
```
This uses the hardware-based sweepers. I assume that the compiled program performs the hardware sweeping in the same shot, then repeats for `num_shots`.

```python
program = program.n_shots(num_shots).sweep(sweep_values, sweep_variables)
```
This uses software-based sweepers. I assume that this repeats the same sequence for `num_shots`, then changes & reprograms the entire sequence for the next sweeper parameter.

Reply:
Found D:

![image](https://github.com/user-attachments/assets/30b1f608-2007-46fa-9528-bf10c94796cb)

Are you aware there are any other limitations of this kind? I.e. calls' order-dependent behaviors.

In case there aren't, a possible way to simplify is relying on the fact that you are already storing the information about this sorting in the `hw_demod` flag, and just push both calls to the end, something like:
```py
sweep = lambda program, sweepers: program.sweep(*sweepers)
program = reduce(sw_sweeps, sweep, reduce(hw_sweeps, sweep, program).n_shots(num_shots))
```
having accumulated in advance the `sw_sweeps` and `hw_sweeps` (whose members will be the `(sweep_values, sweep_variables)` pairs).

I guess they are actually manipulating an instruction stack internally, and the method calls are just pushing more instructions on the stack.

Though, apparently, the instructions for channels and execution could be mixed, as you're first invoking the hardware sweeps, then pushing the waveforms and operations, and eventually defining the shots and software sweeps.

Reply:
> Are you aware there are any other limitations of this kind? I.e. calls' order-dependent behaviors.

Hm, this is the only one I've encountered so far in testing.

> In case there aren't, a possible way to simplify is relying on the fact that you are already storing the information about this sorting in the `hw_demod` flag, and just push both calls to the end, something like:
> 
> ```python
> sweep = lambda program, sweepers: program.sweep(*sweepers)
> program = reduce(sw_sweeps, sweep, reduce(hw_sweeps, sweep, program).n_shots(num_shots))
> ```
> 
> having accumulated in advance the `sw_sweeps` and `hw_sweeps` (whose members will be the `(sweep_values, sweep_variables)` pairs).

Yeah, I will make a modification to handle this better.

Reply:
Unfortunately, nested hardware sweepers are not supported, but we can futureproof it.

Implemented in https://github.com/qiboteam/qibolab/pull/944/commits/1fec5188736b762a23dbdbb33937ec76233076c2